src/api/dirmonitor/polling.c

```diff
@@ -0,0 +1,123 @@
+#include <stdlib.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <SDL.h>
+
+enum {
+  EVENT_MODIFIED
+};
+
+struct file_info {
+  int fd;
+  struct stat file_stat;
+  int events;
+};
+
+struct dirmonitor_internal {
+  struct file_info* fds;
+  size_t fd_count;
+};
+
+  
+struct dirmonitor_internal* init_dirmonitor() {
+  struct dirmonitor_internal* monitor = calloc(1, sizeof(struct dirmonitor_internal));
+
+  monitor->fd_count = 0;
+  monitor->fds = NULL;
+  return monitor;
+}
+
+void deinit_dirmonitor(struct dirmonitor_internal* monitor) {
+  for (size_t i = 0; i < monitor->fd_count; ++i)
+    close(monitor->fds[i].fd);
+  free(monitor->fds);
+}
+
+int get_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int len)
+{
+  int c = 0;
+  for (size_t i = 0; i < monitor->fd_count; ++i) {
+    if (monitor->fds[i].fd < 0)
+      continue;
+
+    struct stat new_stat;
+    fstat(monitor->fds[i].fd, &new_stat);
+
+    if (new_stat.st_mtime > monitor->fds[i].file_stat.st_mtime)
+    {
+      monitor->fds[i].file_stat = new_stat;
+      monitor->fds[i].events = EVENT_MODIFIED;
+      c += 1;
+    }
+  }
+
+  if (!c)
+    SDL_Delay(100);
+
+  return c;
+}
+
+int translate_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int count, int (*change_callback)(int, const char*, void*), void* data) {
+  for (size_t i = 0; i < monitor->fd_count && count; ++i) {
+    if (!(monitor->fds[i].events & EVENT_MODIFIED)) {
+      --count;
+      monitor->fds[i].events = 0;
+      change_callback(monitor->fds[i].fd, NULL, data);
+    }
+  }
+
+  return count != 0;
+}
+
+int add_dirmonitor(struct dirmonitor_internal* monitor, const char* path)
+{
+  int fd = open(path, O_RDONLY | O_NOFOLLOW);
+  if (fd == -1)
+    return -1;
+
+  // Check if there is any space we can reuse
+  for (size_t i = 0; i < monitor->fd_count; ++i) {
+    if (monitor->fds[i].fd == -1)
+    {
+      monitor->fds[i].fd = fd;
+      fstat(fd, &monitor->fds[i].file_stat);
+      return fd;
+    }
+  }
+
+  // There is no empty space, add a new entry;
+  monitor->fd_count++;
+  struct file_info* old_fds = monitor->fds;
+  struct file_info* new_fds;
+  new_fds = malloc(sizeof(*new_fds) * monitor->fd_count);
+  memset(new_fds, 0, sizeof(*new_fds) * monitor->fd_count);
+  if (old_fds)
+    memcpy(new_fds, old_fds, sizeof(*new_fds) * (monitor->fd_count-1));
+
+  new_fds[monitor->fd_count-1].fd = fd;
+  fstat(fd, &new_fds[monitor->fd_count-1].file_stat);
+
+  monitor->fds = new_fds;
+  free(old_fds);
+
+  return fd;
+}
+
+void remove_dirmonitor(struct dirmonitor_internal* monitor, int fd)
+{
+  for (size_t i = 0; i < monitor->fd_count; ++i) {
+    if (monitor->fds[i].fd == fd) {
+      int fd = monitor->fds[i].fd;
+
+      monitor->fds[i].fd = -1;
+      monitor->fds[i].events = 0;
+      close(fd);
+    }
+  }
```

Comment:
Are there cases where multiple `fds` are closed?

Reply:
this interface only allows 1 fd to be closed and `open` won't reuse fd's AFAIK

Reply:
So it would be fine to exit early?

Reply:
ye